File: static/js/test-animation.js
// Test Animation functionality
class TestAnimation {
    constructor() {
        this.testPhrases = [
            "Hello, how are you today?",
            "Welcome to Kan-guroo!",
            "I can help you with educational programs.",
            "What would you like to know?",
            "Let me tell you about our exchange programs.",
            "We offer amazing opportunities for students.",
            "Contact us for more information.",
            "Thank you for your interest!"
        ];
        
        this.visemeMap = {
            'A': 'viseme_aa', 'B': 'viseme_PP', 'C': 'viseme_kk', 'D': 'viseme_DD',
            'E': 'viseme_ee', 'F': 'viseme_ff', 'G': 'viseme_kk', 'H': 'viseme_aa',
            'I': 'viseme_ii', 'J': 'viseme_DD', 'K': 'viseme_kk', 'L': 'viseme_nn',
            'M': 'viseme_PP', 'N': 'viseme_nn', 'O': 'viseme_oo', 'P': 'viseme_PP',
            'Q': 'viseme_kk', 'R': 'viseme_rr', 'S': 'viseme_ss', 'T': 'viseme_DD',
            'U': 'viseme_uu', 'V': 'viseme_ff', 'W': 'viseme_aa', 'X': 'viseme_kk',
            'Y': 'viseme_ii', 'Z': 'viseme_ss'
        };
        
        this.init();
    }
    
    init() {
        this.createTestButton();
        this.setupEventListeners();
    }
    
    createTestButton() {
        // Create test button
        const testButton = document.createElement('button');
        testButton.id = 'testAnimation';
        testButton.className = 'control-button';
        testButton.title = 'Test animation';
        testButton.innerHTML = `
            <svg width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2">
                <path d="M8 5v14l11-7z"></path>
            </svg>
        `;
        
        // Add to character controls
        const controls = document.querySelector('.character-controls');
        if (controls) {
            controls.appendChild(testButton);
        }
        
        // Add test panel
        this.createTestPanel();
    }
    
    createTestPanel() {
        const testPanel = document.createElement('div');
        testPanel.id = 'testPanel';
        testPanel.style.cssText = `
            position: fixed;
            top: 10px;
            left: 10px;
            width: 300px;
            background: rgba(0, 0, 0, 0.9);
            color: white;
            padding: 15px;
            border-radius: 8px;
            z-index: 1000;
            font-family: monospace;
            font-size: 12px;
            display: none;
        `;
        
        testPanel.innerHTML = `
            <h3 style="margin: 0 0 10px 0; color: #FA8148;">🧪 Animation Test Panel</h3>
            <div>
                <label>Test Phrase:</label><br>
                <select id="testPhraseSelect" style="width: 100%; margin: 5px 0;">
                    <option value="custom">Custom text...</option>
                </select>
                <input type="text" id="customPhrase" placeholder="Enter custom text..." style="width: 100%; margin: 5px 0; padding: 5px;">
            </div>
            <div>
                <label>Animation Type:</label><br>
                <select id="animationType" style="width: 100%; margin: 5px 0;">
                    <option value="normal">Normal</option>
                    <option value="exaggerated">Exaggerated</option>
                    <option value="slow">Slow</option>
                    <option value="fast">Fast</option>
                </select>
            </div>
            <div style="margin: 10px 0;">
                <button id="runTest" style="background: #FA8148; color: white; border: none; padding: 8px 15px; border-radius: 4px; cursor: pointer; margin-right: 5px;">Run Test</button>
                <button id="closePanel" style="background: #666; color: white; border: none; padding: 8px 15px; border-radius: 4px; cursor: pointer;">Close</button>
            </div>
            <div id="testResults" style="margin-top: 10px; font-size: 10px;"></div>
        `;
        
        document.body.appendChild(testPanel);
        
        // Populate test phrases
        const select = testPanel.querySelector('#testPhraseSelect');
        this.testPhrases.forEach((phrase, index) => {
            const option = document.createElement('option');
            option.value = phrase;
            option.textContent = phrase;
            select.appendChild(option);
        });
    }
    
    setupEventListeners() {
        // Test button click
        document.addEventListener('click', (e) => {
            if (e.target.closest('#testAnimation')) {
                this.toggleTestPanel();
            }
        });
        
        // Run test button
        document.addEventListener('click', (e) => {
            if (e.target.id === 'runTest') {
                this.runAnimationTest();
            }
        });
        
        // Close panel
        document.addEventListener('click', (e) => {
            if (e.target.id === 'closePanel') {
                this.closeTestPanel();
            }
        });
        
        // Custom phrase input
        document.addEventListener('input', (e) => {
            if (e.target.id === 'customPhrase') {
                const select = document.getElementById('testPhraseSelect');
                if (e.target.value) {
                    select.value = 'custom';
                }
            }
        });
    }
    
    toggleTestPanel() {
        const panel = document.getElementById('testPanel');
        panel.style.display = panel.style.display === 'none' ? 'block' : 'none';
    }
    
    closeTestPanel() {
        const panel = document.getElementById('testPanel');
        panel.style.display = 'none';
    }
    
    runAnimationTest() {
        const select = document.getElementById('testPhraseSelect');
        const customInput = document.getElementById('customPhrase');
        const animationType = document.getElementById('animationType').value;
        const results = document.getElementById('testResults');
        
        let testText = select.value;
        if (testText === 'custom') {
            testText = customInput.value;
        }
        
        if (!testText.trim()) {
            results.innerHTML = '<span style="color: #ff6b6b;">Please enter test text!</span>';
            return;
        }
        
        console.log('🧪 Running animation test with:', testText);
        results.innerHTML = `<span style="color: #4ecdc4;">Testing: "${testText}"</span>`;
        
        // Generate viseme data
        const visemeData = this.generateVisemeData(testText, animationType);
        
        // Start lip-sync animation
        if (window.lipSyncManager) {
            window.lipSyncManager.startLipSync(visemeData);
            results.innerHTML += '<br><span style="color: #45b7d1;">✅ Animation started!</span>';
        } else {
            results.innerHTML += '<br><span style="color: #ff6b6b;">❌ Lip-sync manager not available</span>';
        }
    }
    
    generateVisemeData(text, animationType = 'normal') {
        console.log(`🎭 Generating ${animationType} viseme data for: "${text}"`);
        
        const visemes = [];
        let currentTime = 0.0;
        let durationPerChar = 0.12; // Default 120ms per character
        
        // Adjust timing based on animation type
        switch (animationType) {
            case 'slow':
                durationPerChar = 0.2;
                break;
            case 'fast':
                durationPerChar = 0.08;
                break;
            case 'exaggerated':
                durationPerChar = 0.15;
                break;
        }
        
        for (const char of text.toUpperCase()) {
            if (char in this.visemeMap) {
                const viseme = {
                    time: currentTime,
                    viseme: this.visemeMap[char]
                };
                visemes.push(viseme);
                console.log(`  ${char} -> ${viseme.viseme} at ${viseme.time.toFixed(2)}s`);
            }
            currentTime += durationPerChar;
        }
        
        // Add silence at the end
        visemes.push({
            time: currentTime,
            viseme: 'viseme_sil'
        });
        
        const result = {
            visemes: visemes,
            duration: currentTime,
            text: text,
            test_mode: true,
            animation_type: animationType
        };
        
        console.log(`✅ Generated ${visemes.length} visemes over ${currentTime.toFixed(2)}s`);
        return result;
    }
}

// Initialize test animation when DOM is loaded
document.addEventListener('DOMContentLoaded', () => {
    window.testAnimation = new TestAnimation();
});

// Export for global access
window.TestAnimation = TestAnimation;

File: test_animation.py
#!/usr/bin/env python3
"""
Test file for animation logic - generates test viseme data
"""

import json
import time
import random

class AnimationTester:
    def __init__(self):
        self.test_phrases = [
            "Hello, how are you today?",
            "Welcome to Kan-guroo!",
            "I can help you with educational programs.",
            "What would you like to know?",
            "Let me tell you about our exchange programs.",
            "We offer amazing opportunities for students.",
            "Contact us for more information.",
            "Thank you for your interest!"
        ]
        
        # Viseme mapping for different sounds
        self.viseme_map = {
            'A': 'viseme_aa', 'B': 'viseme_PP', 'C': 'viseme_kk', 'D': 'viseme_DD',
            'E': 'viseme_ee', 'F': 'viseme_ff', 'G': 'viseme_kk', 'H': 'viseme_aa',
            'I': 'viseme_ii', 'J': 'viseme_DD', 'K': 'viseme_kk', 'L': 'viseme_nn',
            'M': 'viseme_PP', 'N': 'viseme_nn', 'O': 'viseme_oo', 'P': 'viseme_PP',
            'Q': 'viseme_kk', 'R': 'viseme_rr', 'S': 'viseme_ss', 'T': 'viseme_DD',
            'U': 'viseme_uu', 'V': 'viseme_ff', 'W': 'viseme_aa', 'X': 'viseme_kk',
            'Y': 'viseme_ii', 'Z': 'viseme_ss'
        }
    
    def generate_test_viseme_data(self, text="Hello, this is a test!"):
        """Generate test viseme data for animation testing"""
        print(f"🎭 Generating test viseme data for: '{text}'")
        
        visemes = []
        current_time = 0.0
        duration_per_char = 0.12  # 120ms per character
        
        for char in text.upper():
            if char in self.viseme_map:
                viseme = {
                    'time': current_time,
                    'viseme': self.viseme_map[char]
                }
                visemes.append(viseme)
                print(f"  {char} -> {viseme['viseme']} at {viseme['time']:.2f}s")
            current_time += duration_per_char
        
        # Add silence at the end
        visemes.append({
            'time': current_time,
            'viseme': 'viseme_sil'
        })
        
        result = {
            'visemes': visemes,
            'duration': current_time,
            'text': text,
            'test_mode': True
        }
        
        print(f"✅ Generated {len(visemes)} visemes over {current_time:.2f}s")
        return result
    
    def generate_exaggerated_viseme_data(self, text="Hello, this is a test!"):
        """Generate exaggerated viseme data for dramatic animation"""
        print(f"🎭 Generating EXAGGERATED viseme data for: '{text}'")
        
        visemes = []
        current_time = 0.0
        duration_per_char = 0.15  # Slower for more dramatic effect
        
        for char in text.upper():
            if char in self.viseme_map:
                viseme = {
                    'time': current_time,
                    'viseme': self.viseme_map[char]
                }
                visemes.append(viseme)
                print(f"  {char} -> {viseme['viseme']} at {viseme['time']:.2f}s")
            current_time += duration_per_char
        
        # Add dramatic pause
        visemes.append({
            'time': current_time,
            'viseme': 'viseme_sil'
        })
        
        result = {
            'visemes': visemes,
            'duration': current_time,
            'text': text,
            'test_mode': True,
            'exaggerated': True
        }
        
        print(f"✅ Generated {len(visemes)} EXAGGERATED visemes over {current_time:.2f}s")
        return result
    
    def generate_random_test_phrase(self):
        """Generate a random test phrase"""
        return random.choice(self.test_phrases)
    
    def test_viseme_frequency(self, text):
        """Analyze viseme frequency in text"""
        viseme_count = {}
        for char in text.upper():
            if char in self.viseme_map:
                viseme = self.viseme_map[char]
                viseme_count[viseme] = viseme_count.get(viseme, 0) + 1
        
        print(f"📊 Viseme frequency analysis for '{text}':")
        for viseme, count in sorted(viseme_count.items()):
            print(f"  {viseme}: {count} times")
        
        return viseme_count

def main():
    """Test the animation logic"""
    print("🧪 ANIMATION TESTER")
    print("=" * 50)
    
    tester = AnimationTester()
    
    # Test 1: Basic viseme generation
    print("\n🎬 Test 1: Basic Viseme Generation")
    test_text = "Hello World!"
    viseme_data = tester.generate_test_viseme_data(test_text)
    
    # Test 2: Exaggerated viseme generation
    print("\n🎭 Test 2: Exaggerated Viseme Generation")
    exaggerated_data = tester.generate_exaggerated_viseme_data(test_text)
    
    # Test 3: Frequency analysis
    print("\n📊 Test 3: Frequency Analysis")
    tester.test_viseme_frequency(test_text)
    
    # Test 4: Random phrase
    print("\n🎲 Test 4: Random Phrase")
    random_phrase = tester.generate_random_test_phrase()
    print(f"Random phrase: '{random_phrase}")
    random_viseme_data = tester.generate_test_viseme_data(random_phrase)
    
    print("\n✅ All tests completed!")
    return viseme_data

if __name__ == '__main__':
    main()
